components/node-labeler/cmd/run.go

```diff
@@ -249,6 +262,150 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req reconcile.Request) (r
+			return ctrl.Result{}, err
+		}
+		// If workspace not found, do a full reconciliation
+		log.WithField("workspace", req.NamespacedName).Info("Workspace not found, reconciling all nodes")
+		return wc.reconcileAllNodes(ctx)
+	}
+
+	if ws.Status.Runtime != nil && ws.Status.Runtime.NodeName != "" {
+		var workspaceList workspacev1.WorkspaceList
+		if err := wc.List(ctx, &workspaceList); err != nil {
+			log.WithError(err).Error("failed to list workspaces")
+			return ctrl.Result{}, err
+		}
+
+		count := 0
+		nodeName := ws.Status.Runtime.NodeName
+		for _, ws := range workspaceList.Items {
+			if ws.Status.Runtime != nil &&
+				ws.Status.Runtime.NodeName == nodeName &&
+				ws.DeletionTimestamp.IsZero() {
+				count++
+			}
+		}
+
+		if err := wc.updateNodeAnnotation(ctx, nodeName, count); err != nil {
+			return ctrl.Result{}, err
+		}
+		log.WithField("node", nodeName).WithField("count", count).Info("updated node annotation")
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (wc *WorkspaceCountController) reconcileAllNodes(ctx context.Context) (ctrl.Result, error) {
+	var workspaceList workspacev1.WorkspaceList
+	if err := wc.List(ctx, &workspaceList); err != nil {
+		log.WithError(err).Error("failed to list workspaces")
+		return ctrl.Result{}, err
+	}
+
+	workspaceCounts := make(map[string]int)
+	for _, ws := range workspaceList.Items {
+		if ws.Status.Runtime != nil &&
+			ws.Status.Runtime.NodeName != "" &&
+			ws.DeletionTimestamp.IsZero() {
+			workspaceCounts[ws.Status.Runtime.NodeName]++
+		}
+	}
+
+	var nodes corev1.NodeList
+	if err := wc.List(ctx, &nodes); err != nil {
+		log.WithError(err).Error("failed to list nodes")
+		return ctrl.Result{}, err
+	}
+
+	for _, node := range nodes.Items {
+		count := workspaceCounts[node.Name]
+		if err := wc.updateNodeAnnotation(ctx, node.Name, count); err != nil {
+			log.WithError(err).WithField("node", node.Name).Error("failed to update node")
+			continue
+		}
+		log.WithField("node", node.Name).WithField("count", count).Info("updated node annotation")
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (wc *WorkspaceCountController) updateNodeAnnotation(ctx context.Context, nodeName string, count int) error {
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		var node corev1.Node
+		err := wc.Get(ctx, types.NamespacedName{Name: nodeName}, &node)
+		if err != nil {
+			return fmt.Errorf("obtaining node %s: %w", nodeName, err)
+		}
+
+		if node.Annotations == nil {
+			node.Annotations = make(map[string]string)
+		}
+
+		if count > 0 {
+			node.Annotations["cluster-autoscaler.kubernetes.io/scale-down-disabled"] = "true"
+			log.WithField("nodeName", nodeName).Info("disabling scale-down for node")
+		} else {
+			delete(node.Annotations, "cluster-autoscaler.kubernetes.io/scale-down-disabled")
+			log.WithField("nodeName", nodeName).Info("enabling scale-down for node")
+		}
+
+		return wc.Update(ctx, &node)
```

Comment:
Observations:
1. This doesn't timeout, but the below `updateLabel` function does. It would probably make sense to do similar.
   * this is blocking to me
2. The below `updateLabel` receives and uses a `client` object to act on k8s. You're using the `get` on the new controller.
   * this is cool, I'm just sharing...that it's cool